fv3core/grid/geometry.py

```diff
@@ -0,0 +1,905 @@
+        sg_field_in[::-1, :], sg_field_out[::-1, :], nhalo
+    )
+
+
+def _rotate_trig_sg_ne_counterclockwise(sg_field_in, sg_field_out, nhalo):
+    _rotate_trig_sg_sw_counterclockwise(
+        sg_field_in[::-1, ::-1], sg_field_out[::-1, ::-1], nhalo
+    )
+
+
+def _rotate_trig_sg_ne_clockwise(sg_field_in, sg_field_out, nhalo):
+    _rotate_trig_sg_sw_clockwise(
+        sg_field_in[::-1, ::-1], sg_field_out[::-1, ::-1], nhalo
+    )
+
+
+def calculate_divg_del6(
+    sin_sg, sina_u, sina_v, dx, dy, dxc, dyc, nhalo, tile_partitioner, rank
+):
+
+    divg_u = sina_v * dyc / dx
+    del6_u = sina_v * dx / dyc
+    divg_v = sina_u * dxc / dy
+    del6_v = sina_u * dy / dxc
+
+    if tile_partitioner.on_tile_bottom(rank):
+        divg_u[:, nhalo] = (
+            0.5
+            * (sin_sg[:, nhalo, 1] + sin_sg[:, nhalo - 1, 3])
+            * dyc[:, nhalo]
+            / dx[:, nhalo]
+        )
+        del6_u[:, nhalo] = (
+            0.5
+            * (sin_sg[:, nhalo, 1] + sin_sg[:, nhalo - 1, 3])
+            * dx[:, nhalo]
+            / dyc[:, nhalo]
+        )
+    if tile_partitioner.on_tile_top(rank):
+        divg_u[:, -nhalo - 1] = (
+            0.5
+            * (sin_sg[:, -nhalo, 1] + sin_sg[:, -nhalo - 1, 3])
+            * dyc[:, -nhalo - 1]
+            / dx[:, -nhalo - 1]
+        )
+        del6_u[:, -nhalo - 1] = (
+            0.5
+            * (sin_sg[:, -nhalo, 1] + sin_sg[:, -nhalo - 1, 3])
+            * dx[:, -nhalo - 1]
+            / dyc[:, -nhalo - 1]
+        )
+    if tile_partitioner.on_tile_left(rank):
+        divg_v[nhalo, :] = (
+            0.5
+            * (sin_sg[nhalo, :, 0] + sin_sg[nhalo - 1, :, 2])
+            * dxc[nhalo, :]
+            / dy[nhalo, :]
+        )
+        del6_v[nhalo, :] = (
+            0.5
+            * (sin_sg[nhalo, :, 0] + sin_sg[nhalo - 1, :, 2])
+            * dy[nhalo, :]
+            / dxc[nhalo, :]
+        )
+    if tile_partitioner.on_tile_right(rank):
+        divg_v[-nhalo - 1, :] = (
+            0.5
+            * (sin_sg[-nhalo, :, 0] + sin_sg[-nhalo - 1, :, 2])
+            * dxc[-nhalo - 1, :]
+            / dy[-nhalo - 1, :]
+        )
+        del6_v[-nhalo - 1, :] = (
+            0.5
+            * (sin_sg[-nhalo, :, 0] + sin_sg[-nhalo - 1, :, 2])
+            * dy[-nhalo - 1, :]
+            / dxc[-nhalo - 1, :]
+        )
+
+    return divg_u, divg_v, del6_u, del6_v
+
+
+def calculate_grid_z(ec1, ec2, vlon, vlat, np):
+    z11 = np.sum(ec1 * vlon, axis=-1)
+    z12 = np.sum(ec1 * vlat, axis=-1)
+    z21 = np.sum(ec2 * vlon, axis=-1)
+    z22 = np.sum(ec2 * vlat, axis=-1)
+    return z11, z12, z21, z22
+
+
+def calculate_grid_a(z11, z12, z21, z22, sin_sg5):
+    a11 = 0.5 * z22 / sin_sg5
+    a12 = -0.5 * z12 / sin_sg5
+    a21 = -0.5 * z21 / sin_sg5
+    a22 = 0.5 * z11 / sin_sg5
+    return a11, a12, a21, a22
+
+
+def edge_factors(
+    grid_quantity, agrid, grid_type, nhalo, tile_partitioner, rank, radius, np
```

Comment:
For functions with many arguments, can you please add type hints? It would help to have docstrings too, but type hints (including output type(s)) go a long way for documentation.

I believe there's a `NumpyModule` type hint available in `fv3gfs.util.typing` or some other `typing` module.